clients/admin-ui/src/features/common/form/inputs.tsx

```diff
@@ -41,66 +54,208 @@ interface InputProps {
 type StringArrayField = FieldHookConfig<string[] | undefined>;
 
+const Label = ({
+  title,
+  ...labelProps
+}: {
+  title: string;
+} & FormLabelProps) => (
+  <FormLabel size="sm" {...labelProps}>
+    {title}
+  </FormLabel>
+);
+
+const TextInput = ({
+  isPassword,
+  ...props
+}: InputProps & { isPassword: boolean }) => {
+  const [type, setType] = useState<"text" | "password">(
+    isPassword ? "password" : "text"
+  );
+
+  const handleClickReveal = () =>
+    setType(type === "password" ? "text" : "password");
+
+  return (
+    <InputGroup size="sm" mr="2">
+      <Input {...props} type={type} pr={isPassword ? "10" : "3"} />
+      {isPassword ? (
+        <InputRightElement pr="2">
+          <IconButton
+            size="xs"
+            variant="unstyled"
+            aria-label="Reveal/Hide Secret"
+            icon={
+              <EyeIcon
+                boxSize="full"
+                color={type === "password" ? "gray.400" : "gray.700"}
+              />
+            }
+            onClick={handleClickReveal}
+          />
+        </InputRightElement>
+      ) : null}
+    </InputGroup>
+  );
+};
+
+const ErrorMessage = ({
+  isInvalid,
+  message,
+  fieldName,
+}: {
+  isInvalid: boolean;
+  fieldName: string;
+  message?: string;
+}) => {
+  if (!isInvalid) {
+    return null;
+  }
+  return (
+    <FormErrorMessage data-testid={`error-${fieldName}`}>
+      {message}
+    </FormErrorMessage>
+  );
+};
+
+const SelectInput = ({
+  options,
+  fieldName,
+  size,
+  isSearchable,
+  isClearable,
+  isMulti = false,
+}: { fieldName: string; isMulti?: boolean } & Omit<SelectProps, "label">) => {
+  const [initialField] = useField(fieldName);
+  const field = { ...initialField, value: initialField.value ?? [] };
+  const selected = isMulti
+    ? options.filter((o) => field.value.indexOf(o.value) >= 0)
+    : options.find((o) => o.value === field.value) || null;
+
+  // note: for Multiselect we have to do setFieldValue instead of field.onChange
+  // because field.onChange only accepts strings or events right now, not string[]
+  // https://github.com/jaredpalmer/formik/issues/1667
+  const { setFieldValue, touched, setTouched } = useFormikContext();
+
+  const handleChangeMulti = (newValue: MultiValue<Option>) => {
+    setFieldValue(
+      field.name,
+      newValue.map((v) => v.value)
+    );
+  };
+  const handleChangeSingle = (newValue: SingleValue<Option>) => {
+    if (newValue) {
+      field.onChange(fieldName)(newValue.value);
+    } else if (isClearable) {
+      field.onChange(fieldName)("");
+    }
+  };
+
+  const components = isClearable ? undefined : { ClearIndicator: () => null };
+
+  return (
+    <Select
+      options={options}
+      onBlur={(e) => {
+        setTouched({ ...touched, [field.name]: true });
+        field.onBlur(e);
+      }}
+      // @ts-ignore dynamic isMulti makes typing this complicated
+      onChange={isMulti ? handleChangeMulti : handleChangeSingle}
```

Comment:
could not figure out the typescript for this one :(

Reply:
Looks like it's possible to do it with generics and/or overloads, or split out the component call to another function so that `<Select isMulti={false}` and `<Select isMulti={true}` infer the exact type.

But I'd probably go with one of these:

```ts
 // Using `as` slightly better then ts-ignore. Technically unsafe, but only as unsafe
  // as react-select's runtime checking.
  const handleChange = (newValue: MultiValue<Option> | SingleValue<Option>) =>
    isMulti
      ? handleChangeMulti(newValue as MultiValue<Option>)
      : handleChangeSingle(newValue as SingleValue<Option>);

  // Adding type guards that also have runtime checks works. May be redundant.
  const handleChange = (newValue: MultiValue<Option> | SingleValue<Option>) =>
    narrow([{ label: "string", value: "string" }], newValue)
      ? handleChangeMulti(newValue)
      : narrow({ label: "string", value: "string" }, newValue)
      ? handleChangeSingle(newValue)
      : undefined; // Throw error?
```

And then `onChange={handleChange}`